### ISSUE TITLE
After resume from sleep OGCS attempts sync before network becomes available.

### ISSUE THREAD

**Issue by kuraikaze** [phw198/OutlookGoogleCalendarSync#1836]
**OGCS Version**: v2.10.4.0 
**Edition**: Installed
**Sync Direction**: O<->G

**Bug Description**  
After PC resumes from S3 sleep or hibernation OGCS attempts a sync cycle immediately, frequently before the network stack has completed re-initialisation, leading to a connection failure which generates the error popup.

**Steps to Reproduce the Issue**  
1. Sleep/hibernate PC
2. Resume PC.

**Any other information**  
Regression to behaviour from closed issues #886, #1144

:paperclip:  [OGcalsync.log gist](https://gist.github.com/kuraikaze/65c67ad644f4ed23373a3ec4b3494aba)
(Resume event at 2024-04-14 08:54:11,214)

**Comment by phw198** [comment]
Thanks for reporting @kuraikaze 👍🏻 

:package: Hopefully this should now be resolved with [hotfix v2.10.4.2.zip](https://github.com/phw198/OutlookGoogleCalendarSync/files/15051624/v2.10.4.2.zip) ([see instructions for applying](https://github.com/phw198/OutlookGoogleCalendarSync/wiki/Reporting-Problems#applying-a-hotfix)).
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [CRCs](https://github.com/phw198/OutlookGoogleCalendarSync/wiki/Known-Issues#cyclic-redundancy-checks---aka-crcs): Zip = `611B5D18` Exe = `9DFB1E62`

Let me know how it goes.

**Comment by kuraikaze** [comment]
Thanks for the update. *Almost* there with the hotfix...
The app no longer brings up it's window and an error popup upon resume but it still generates a Windows Notification:-
"Issue encountered.
Please review output on the main 'Sync' tab"

The app then successfully completed a sync after a few minutes, before I got a chance to look at the details in the Sync window.
I've turned on Debug logging and I'll include the log next time I do a sleep/resume cycle.

**Comment by kuraikaze** [comment]
Gist of log for Hotfix version.
Resume event at 2024-04-22 19:32:01 : [OGcalsync.log 2.10.4.2-hotfix](https://gist.github.com/kuraikaze/65c67ad644f4ed23373a3ec4b3494aba#file-ogcalsync-log-2-10-4-2-hotfix)

**Comment by phw198** [comment]
Thanks @kuraikaze, I think that looks correct and I'm sure that was its behaviour before v2.10.4 - it should report the error (but only in the sync window and not as a popup message box) and automatically retry within 15 minutes.

**Comment by kuraikaze** [comment]
Yep - the only difference now is that the initial error at resume time generates a Windows Notification, which didn't happen prior to V2.10.4

**Comment by phw198** [comment]
That's more of an intentional change to always have a system notification when there are errors. However, as per #1840 it probable needs some refinement with maybe 4 options to chose from: no notifications / only errors / errors + changes / all.

**Comment by kuraikaze** [comment]
Something like that would work for me. Or, perhaps "Notify only after 2 consecutive errors"
Alternatively, could there be a mechanism to delay the first sync after resume for, say, 15 seconds ?
I'm not familiar enough with the internals of Windows to know if there's any easily accessible method to inform userland applications of the S3 resume event.

**Comment by phw198** [comment]
@kuraikaze I've created #1850 to see if we can delay syncs immediately after resume - please have a go with the hotfix there.

**Comment by kuraikaze** [comment]
Hi Paul
This issue seems to have returned with v2.11.4-Alpha
Would [hotfix v2.11.4.1.zip](https://github.com/user-attachments/files/20104636/v2.11.4.1.zip) from https://github.com/phw198/OutlookGoogleCalendarSync/issues/2117 also be relevant for this ?

Log attached:
[[OGcalsync.log gist](https://gist.github.com/kuraikaze/7d0076527a521eb8b6caead33ef47b03)]

**Comment by phw198** [comment]
No I don't think it has returned, though if it had it would be related to #1850, as that was putting in a delay for an auto-sync following resume.

However, as I noted in that issue "OGCS already detected a system time change, which is a side effect of resuming from suspend and logged at INFO with "Detected system timezone change" - so I think we can piggyback this."

But in your log just posted:-
```
2025-05-21 08:36:48,502 DEBUG 1  OutlookGoogleCalendarSync.Sync.SyncTimer [54] -  Scheduled sync triggered for profile: Default
2025-05-21 08:36:49,504 INFO  12 OutlookGoogleCalendarSync.NotificationTray [50] -  Detected system timezone change.
```
So the event OGCS is using fired a second _after_ the sync was triggered, so the sync proceeded as though there had been no resume event.

Given it sounds like this is the first time this has happened for you in a year, I think I'll just put this down to bad luck (coding in proper detection of resume from hibernate etc across various versions of Windows looks like a complete nightmare and I wouldn't be convinced that would fire before the scheduled sync again!)... 🙁

**Comment by kuraikaze** [comment]
It wasn't an isolated one-off ... it happens on every sleep/resume longer than 1 hour with v2.11.4-Alpha

Given, as you say, that consistent detection of Windows events is a nightmare to navigate I can't rule out that there's something under the Windows hood that has changed, rather than OGCS.

To verify I'll revert to v2.11.2-Alpha/v2.11.3-Alpha and see if the behaviour goes away.

**Comment by phw198** [comment]
Oh OK! I'll be rather surprised if the issue isn't reproducible on previous alphas, but you never know so certainly worth confirming.

I'm happy to try adding an event to detect S3 sleep as opposed to resume (though this won't work if you or anyone else is using [Modern Standby](https://learn.microsoft.com/en-us/windows-hardware/design/device-experiences/modern-standby) on Windows 10 or later) - just let me know once you've tried reverting to previous alphas...

**Comment by kuraikaze** [comment]
Mystery ~mostly~ solved :thumbsup:
Not an issue with recent code commits/releases at all.

When I went to change the in-place .4-Alpha for .2 or .3 I discovered, in passing, that I had two instances of OGCS running - this seemed to cause all sorts of weirdness.

e.g. (not displaying in tray, some of the OGCS tabs not rendering or interactable when I manually brought it's window to fore (via Sysinternals ProcessExplorer), not applying the sync delay after resume causing error notification etc.)
This was probably caused by the running instance losing  it's tray/taskbar hook during one of those annoying Windows explorer wobbles during monitor resolution/priority switching that some fullscreen games trigger.

When I didn't see OGCS in tray after running the .4-Alpha installer I ran the \AppData\Local\OutlookGoogleCalendarSync\app-2.11.4-alpha\OutlookGoogleCalendarSync.exe /d:180 cmdline manually.

~The second instance which *did* show in the tray, would seem to behave normally, and could write changes to the settings file that the 1st instance would then implement, but it was the b0rked 1st instance that would lock & write to the log file and so on.~
_Above line made irrelevant by subsequent event_

After manually killing all instances and removing/re-adding the HKCU\...\CurrentVersion\Run and a reboot I now have a working, single instance, .4-Alpha install, and the resume/sync-delay mechanism works as intended.

Sorry for the confusion. :blush:

~As a remedy to forestall future user-instigated SNAFUs in this area could I request a feature addition to only allow a single running~
Never mind - I manually ran a bunch of extra instances to try and reproduce and *just* as I was typing the above line the "Multiple OGCS instances running" modal warning popup triggered.  That's me good and told off :)

**Comment by kuraikaze** [comment]
Welp ... scratch all that.  A properly installed and solo instance of v2.11.4-Alpha fails to respect the wait period after a wake from sleep event.

Going back to the earlier plan of testing earlier betas as I can't recall exactly when this behaviour crept back in.

**Update:**
.3-Alpha also fails to respect the wait period.
.2-Alpha test in progress - waiting for a >1 hour sleep so that the "system timezone change" gets activated.

**Comment by kuraikaze** [comment]
Looks like it is indeed something that's changed in the way this PC and Windows 10 enters and exits sleep and/or how it brings up the network stack..
I tried all the 2.11.x-Alpha builds, the earliest of which certainly precede my hazy recall of when this issue re-occured for me, (anywhere between 1 and 2 months).  All the builds now show the initial S3 resume error behaviour, which they certainly didn't before.

So, after all that waffle ...

Yes, I'd be very happy if you would investigate a way to add an event to detect S3 sleep as opposed to resume, thank you.  :thumbsup:

**Comment by phw198** [comment]
Gah, I was hoping to avoid that! 😆   But thanks for doing the legwork to confirm the issue isn't directly OGCS.

Have now raised #2132 to track that enhancement on.
